feat/emote-menu-sorting: Add sorting to emote menu.

berghall opened the pull request:
The sorting will default to newest added emote.

Closes #398, #511

AnatoleAM commented on src/site/twitch.tv/modules/emote-menu/EmoteMenuContext.ts:
```diff
@@ -22,3 +22,16 @@ export function useEmoteMenuContext(): EmoteMenuContext {
+};
+
+export const emoteMenuSortProperties: EmoteMenuSort = {
+	Name: "Name",
+	Listed: "Listed",
+	Animated: "Animated",
+	Timestamp: "Date added",
```
Keys should be lowercase

```suggestion
	name: "Name",
	listed: "Listed",
	animated: "Animated",
	timestamp: "Date added",
```

AnatoleAM commented on src/site/twitch.tv/modules/emote-menu/EmoteMenuModule.vue:
```diff
@@ -115,6 +122,18 @@ export const config = [
 	}),
+	declareConfig<EmoteMenuSortPropertyKey>("ui.emote_menu.sort_emotes_by", "DROPDOWN", {
+		path: ["Appearance", "Interface"],
+		label: "Emote Menu: Sort emotes by property",
```
nit: shorten the label

```suggestion
		label: "Emote Menu: Sorting",
		hint: "Define how emotes are sorted in the Emote Menu",
```

AnatoleAM commented on src/site/twitch.tv/modules/emote-menu/EmoteMenuSet.vue:
```diff
@@ -86,6 +88,30 @@ function sortCase(ae: SevenTV.ActiveEmote): number {
+const getEmoteMenuSortBy = (a: SevenTV.ActiveEmote, b: SevenTV.ActiveEmote): number => {
+	const nameSort = a.name.localeCompare(b.name);
+	switch (sortBy.value) {
+		case "Timestamp": {
+			if (!a.timestamp || !b.timestamp) {
+				return nameSort;
+			}
+			return a.timestamp > b.timestamp ? 1 : -1;
+		}
+		case "Name": {
+			return nameSort;
+		}
+		case "Listed": {
+			return a.data?.listed === b.data?.listed ? 1 : -1;
+		}
+		case "Animated": {
+			return a.data?.animated === b.data?.animated ? 1 : -1;
+		}
```
update keys

```suggestion
		case "timestamp": {
			if (!a.timestamp || !b.timestamp) {
				return nameSort;
			}
			return a.timestamp > b.timestamp ? 1 : -1;
		}
		case "name": {
			return nameSort;
		}
		case "listed": {
			return a.data?.listed === b.data?.listed ? 1 : -1;
		}
		case "animated": {
			return a.data?.animated === b.data?.animated ? 1 : -1;
		}
```

AnatoleAM commented on src/site/twitch.tv/modules/emote-menu/EmoteMenuModule.vue:
```diff
@@ -115,6 +122,18 @@ export const config = [
+		options: emoteSortByProperties,
+		defaultValue: "Timestamp",
+	}),
+	declareConfig<boolean>("ui.emote_menu.sort_emotes_desc", "TOGGLE", {
+		path: ["Appearance", "Interface"],
+		label: "Emote Menu: Sort emotes by descending order",
+		hint: "Sort emotes by descending order, sort by ascending order if off",
+		defaultValue: true,
+	}),
```
This should not be on by default. Also, it could be made easier to understand; change the wording here, with a label such as `Emote Menu: Reverse Sorting Order`.

AnatoleAM reviewed (changes requested):
This currently breaks one core design point of the emote menu, which is width sort, with wider emotes towards the bottom. This should continue to apply regardless of other sorting checks.

AnatoleAM reviewed (commented):
Functionality wise this works fine now. In terms of user experience though, this isn't great.

The sorting options should be within the emote menu itself. i.e, next to the search bar.

berghall commented:
@AnatoleAM I guess I should use the font-awesome icons and I need the svg, so could I get them from you?
Was thinking for the order these two https://fontawesome.com/icons/sort-up?f=classic&s=duotone&sc=%231E3050 and https://fontawesome.com/icons/sort-down?f=classic&s=duotone&sc=%231E3050.
For the sort type https://fontawesome.com/icons/bars-sort?f=classic&s=solid

AnatoleAM commented:
@berghall added the icons on your branch

berghall commented:
@AnatoleAM controls are now next to the search input

![f57da1c9bb6d2dfc54d174c756a4611c](https://user-images.githubusercontent.com/20751720/233459867-8e13d3f8-a347-45df-9bf3-b79248e1d490.png)
![17ab7321fdd1768dd852c3e5c856f50c](https://user-images.githubusercontent.com/20751720/233459878-b94212d6-4ec1-4040-819c-a42241f962c7.png)

Live input search:
![3159dc97f5c7a18ce158fe18578ea920](https://user-images.githubusercontent.com/20751720/233460268-d3e75d9f-049c-4b57-b6c9-123e5e116460.png)

ayyybubu commented:
Not really a fan of adding more clutter to the menu and shrinking the size of the search bar. All of this could be wrapped inside a single context menu. 
![dJdEix6](https://user-images.githubusercontent.com/88165881/233612550-5154e248-f51b-4f09-aef3-ace3ffb8aa31.png)

berghall commented:
If I were to combine them into one menu, then I'd need these two: 
https://fontawesome.com/icons/arrow-up-wide-short?f=classic&s=solid
https://fontawesome.com/icons/arrow-down-wide-short?f=classic&s=solid

ayyybubu commented:
You can just use the bars-sort icon you already have. It doesn't really matter

berghall commented:
Updated
![ba0bfe5215f365b1201689b3673b7713](https://user-images.githubusercontent.com/20751720/234356365-be08e146-6d6b-406f-b8ad-6472a26693ac.png)

Changed the position when live input search is on, to top of settings button
![16dfcf5b9b8d25bf2797b2203f0fe073](https://user-images.githubusercontent.com/20751720/234356160-ea3cf064-5a44-4f11-889c-d42d2f764094.png)

berghall commented:
@ayyybubu your requested changes have been implemented.

ayyybubu commented:
Thank you, it looks pretty clean now. The last thing that i forgot to mention before:
-clicking on the sort button again should hide the active context menu

berghall commented:
> -clicking on the sort button again should hide the active context menu
@ayyybubu this is now fixed.

berghall commented:
@AnatoleAM your requested changes have been implemented.

berghall commented:
@AnatoleAM merge conflicts resolved

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SevenTV/Extension?pullRequest=516) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/SevenTV/Extension?pullRequest=516) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: berghall<br/>:x: AnatoleAM<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SevenTV/Extension?pullRequest=516) it.</sub>

TroyKomodo commented:
Features against this repo wont be implemented, we are going to rewrite the extension on this [repo](https://github.com/SevenTV/SevenTV).
